File: ASD_tv.cxx
// Minimize the Objects TV with ASD - POCS.
// tomo_tv.cxx
//  TV
//
//  Created by Jonathan Schwartz on 5/6/19.
//  Copyright © 2019 Jonathan Schwartz. All rights reserved.
//

#include <iostream>
#include <sys/stat.h>
#include <Eigen/Core>
#include <Eigen/SparseCore>
#include <opencv2/opencv.hpp>
#include <opencv2/core/eigen.hpp>
#include "tlib.cpp"

using namespace Eigen;
using namespace std;
using namespace cv;

///////////////RECONSTRUCTION PARAMETERS///////////////////

//File Name (Input Tilt Series).
String filename = "Co2P_v2.tif";

int Niter, ng;                 //Number of Iterations (Main and TV).
float dTheta;                  //Step Size for Theta.
float beta, beta_red;          //Parameter in ART Reconstruction.
float eps;                     //Data Tolerance Parameter.
float alpha, alpha_red, r_max; //TV Parameter and reduction criteria.

int Nc = 100; // Number of Counts for Poisson Noise. 

///////////////////////////////////////////////////////////

int main(int argc, const char * argv[]) {
    
    //Assign Values from parameters.txt
    read_parameters(Niter, ng, dTheta, beta, beta_red, alpha, alpha_red, eps, r_max);
    
    //Load Dataset.
    Mat img = imread("Test_Images/" + filename, cv::ImreadModes::IMREAD_GRAYSCALE);
    int Nslice = img.rows;
    int Nray = img.cols;
    Eigen::MatrixXf tiltSeries;
    cv::cv2eigen(img, tiltSeries);

    //Generate Measurement Matrix.
    int Nproj = 180/dTheta + 1;
    VectorXf tiltAngles = VectorXf::LinSpaced(Nproj, 0, 180);
    int Nrow = Nray*Nproj;
    int Ncol = Nray*Nray;
    SparseMatrix<float, Eigen::RowMajor> A(Nrow,Ncol);
    parallelRay(Nray, tiltAngles, A);

    //Calculate Inner Product.
    VectorXf rowInnerProduct(Nrow);
    for(int j=0; j < Nrow; j++)
    {
        rowInnerProduct(j) = A.row(j).dot(A.row(j));
    }

    //Vectorize/Initialize the reconstruction and experimental data.
    tiltSeries.resize(tiltSeries.size(), 1);
    VectorXf b = A * tiltSeries;
    //poissonNoise(b, Nc); //Uncoment if you'd like to add poisson Noise.
    tiltSeries.resize(Nslice, Nray);
    
    // Empty Vectors/Matrices for Reconstruction. 
    VectorXf g;
    MatrixXf recon (Nslice, Nray), temp_recon(Nslice, Nray), v;
    recon.setZero();
    float dPOCS;
    
    //Vectors to evalutate convergence.
    VectorXf dd_vec(Niter), dp_vec(Niter), dg_vec(Niter);
    VectorXf dPOCS_vec(Niter), beta_vec(Niter), rmse_vec(Niter);
    VectorXf cos_alpha_vec(Niter), tv_vec(Niter);

    //Main Loop.
    for(int i=0; i < Niter; i++)
    {
        // Print the Iteration Number.
        if ( i % 100 == 0)
            cout << "Iteration: " << i + 1 << " / " << Niter << "\n";
        
        temp_recon = recon;
        beta_vec(i) = beta;
        
        //ART Reconstruction.
        tomography(recon, b, rowInnerProduct, A, beta);
        recon = (recon.array() < 0).select(0, recon);
        g = A * recon;
        recon.resize(Nslice, Nray);
        
        if(i == 0)
        {
            dPOCS = (recon - temp_recon).norm() * alpha;
        }
        
        dd_vec(i) = (g - b).norm() / g.size();
        dp_vec(i) = (temp_recon - recon).norm();
        temp_recon = recon;

        //TV Loop.
        for(int j=0; j<ng; j++)
        {
            v = tv2Dderivative(recon);
            v /= v.norm();
            recon.array() -= dPOCS * v.array();
        }

        dg_vec(i) = (recon - temp_recon).norm();

        //Reduce TV if data constraint isn't met.
        if (dg_vec(i) > dp_vec(i) * r_max && dd_vec(i) > eps)
        {
            dPOCS *= alpha_red;
        }
        
        dPOCS_vec(i) = dPOCS;
        beta *= beta_red;
        rmse_vec(i) = (tiltSeries - recon).norm();
        cos_alpha_vec(i) = CosAlpha(recon, v, g, b, A);
        tv_vec(i) = tv2D(recon);

    }

    //Create Directory to Save Results.
    mkdir("Results/ASD_tv", ACCESSPERMS);
    
    //Save all the vectors.
    saveVec(beta_vec, "beta");
    saveVec(dd_vec, "dd");
    saveVec(dp_vec, "dp");
    saveVec(dg_vec, "dg");
    saveVec(dPOCS_vec, "dPOCS");
    saveVec(rmse_vec, "RMSE");
    saveVec(cos_alpha_vec, "Cos_Alpha");
    saveVec(tv_vec, "TV");

    //Display the final reconstruction.
//    Mat final_img;
//    cv::eigen2cv(recon, final_img);
//    namedWindow( "Reconstruction", WINDOW_AUTOSIZE );
//    imshow( "Reconstruction", final_img * (1.0 / recon.maxCoeff()) );
//    waitKey(0);
    
    //Save the Image.
    recon.resize(Nslice, Nray);
    Mat final_img;
    cv::eigen2cv(recon, final_img);
    final_img /= recon.maxCoeff();
    imwrite("Results/ASD_tv/recon.tif", final_img);
    
    return 0;
}

File: pycTV/3D/ASD_tv.py
# General 3D - ASD/TV Reconstruction with Positivity Constraint. 

import sys
sys.path.append('./Utils')
from pytvlib import parallelRay, timer, load_data
from matplotlib import pyplot as plt
from skimage import io
import numpy as np
import ctvlib 
import time
########################################

file_name = 'Co2P_tiltser.tif'

# Number of Iterations (Main Loop)
Niter = 100

# Number of Iterations (TV Loop)
ng = 10

# Parameter in ART Reconstruction.
beta = 1.0

# ART Reduction.
beta_red = 0.975

# Data Tolerance Parameter
eps = 0.6

# Reduction Criteria
r_max = 0.95
alpha_red = 0.95
alpha = 0.5

#Show final results (i.e. tv and dd)
show = True

# Save final Reconstruction. 
save = True

##########################################

# #Read Image. 
(file_name, tiltSeries) = load_data(file_name)
(Nslice, Nray, Nproj) = tiltSeries.shape
b = np.zeros([Nslice, Nray*Nproj])

# Initialize C++ Object.. 
tomo_obj = ctvlib.ctvlib(Nslice, Nray, Nproj)

for s in range(Nslice):
    b[s,:] = tiltSeries[s,:,:].transpose().ravel()
tomo_obj.setTiltSeries(b)
tiltSeries = None

# Generate Tilt Angles.
tiltAngles = np.load('Tilt_Series/'+ file_name +'_tiltAngles.npy')

# Generate measurement matrix
A = parallelRay(Nray, tiltAngles)
tomo_obj.load_A(A)
A = None
tomo_obj.rowInnerProduct()

dd_vec = np.zeros(Niter)
# tv_vec = np.zeros(Niter)

t0 = time.time()
counter = 1

#Main Loop
for i in range(Niter): 

    if (i%10 ==0):
        print('Iteration No.: ' + str(i+1) +'/'+str(Niter))

    tomo_obj.copy_recon()

    #ART Reconstruction. 
    tomo_obj.ART(beta, -1)

    #Positivity constraint 
    tomo_obj.positivity()

    #ART-Beta Reduction
    beta *= beta_red 

    #Forward Projection.
    tomo_obj.forwardProjection(-1)

    #Measure Magnitude for TV - GD.
    if (Niter[i] == 0):
        dPOCS = tomo_obj.matrix_2norm() * alpha
        dp = dPOCS / alpha
    else: # Measure change from ART.
        dp = tomo_obj.matrix_2norm() 

    # Measure difference between exp/sim projections.
    dd_vec[i] = tomo_obj.vector_2norm()
    tomo_obj.copy_recon() 

    #TV Minimization. 
    tomo_obj.tv_gd(ng, dPOCS)
    dg = tomo_obj.matrix_2norm()

    if(dg > dp * r_max and dd_vec[i] > eps):
        dPOCS *= alpha_red

    if (i%10 ==0):
        timer(t0, counter, Niter)
    counter += 1

#Garbage collector (gc)
tomo_obj.release_memory()

#Get the final reconstruction. 
recon = np.zeros([Nslice, Nray, Nray], dtype=np.float32, order='F') 
for s in range(Nslice):
    recon[s,:,:] = tomo_obj.getRecon(s)

if show:
    #Create function for plotting data. 

if save:
    np.save('Results/TV_'+ file_name + '_recon.npy', recon)